nixos/steam: add localNetworkGameTransfers.openFirewall option

opl- opened the pull request:
Steam local network game transfers require TCP port 27040 to be open: https://steamcommunity.com/groups/SteamClientBeta/discussions/0/3775742015034590856/#c3827537203130812127

They also appear to use UDP port 27036, which I discovered after setting `remotePlay.openFirewall = false;` broke local transfers, and checked using Wireshark. I saw no traffic on any of the other ports from the 27031-27036 UDP port range used by remote play.

(The [support article](https://help.steampowered.com/en/faqs/view/46BD-6BA8-B012-CE43) mentions port 24070, but the log file `logs\content_log.txt` mentioned in the discussion above supports that it's port 27040: `[2024-02-01 21:04:40] Peer - hostname (192.168.0.2:27040): Closing connection (594 chunks requested, 1069 received = 87.11 MB)`)

Closes #238305

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Adds option `programs.steam.localNetworkGameTransfers.openFirewall`, which opens the appropriate ports.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Tested by adding `networking.firewall.allowedTCPPorts = [ 27040 ]; networking.firewall.allowedUDPPorts = [ 27036 ];` to my system configuration.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

opl- commented:
Marked as draft because I'm not absolutely certain it works. @mkg20001, as the module maintainer, would you be able to test this?

opl- commented:
Seems GitHub search failed me here.

They also named the option slightly differently, including the `Game`. I probably should include that word too.

I also noticed that since I read the Steam Community discussion, someone else mentioned port 27036 being the missing ingredient: https://steamcommunity.com/groups/SteamClientBeta/discussions/0/3775742015034590856/?ctp=2#c4038104984939951535

MinerSebas commented:
In my Testing, the only time a local Transfer happened was when setting:
```
networking.firewall = {
  allowedTCPPorts = [ 27036 27040 ];
  allowedUDPPorts = [ 27036 ];
};
```
In my setup I didn't activate `remotePlay.openFirewall`, which also adds 27036 to the TCP Ports.

To me this implies that `localNetworkTransfers.openFirewall` should also activate `remotePlay.openFirewall` (or throw an assertion if `remotePlay.openFirewall` isn't true).

opl- commented:
I asked Valve's martino, and they said that the feature uses TCP port 27040, and UDP ports 27031-27036; it also seems they updated the Steam help article accordingly.

In practice, using WireShark, I've only seen Steam use ports TCP 27040 (for game data transfers) and UDP 27036 (for client discovery). UDP ports 27031-27035 never saw any traffic.

I asked martino about whether ports 27031-27035 are actually used, given that we're setting up firewall rules here and wanna be as strict as possible, but they did not get back to me on that yet.

@MinerSebas, I did not need TCP port 27036 to be open - just UDP. Are you sure that's not the case for you?

The [Required ports for Steam](https://help.steampowered.com/en/faqs/view/2EA8-4D75-DA21-31EB) help article mentions that UDP port 27036 is used for Steam Remote Play, and Wireshark seems to think data on port 27036 is used for "Steam In House Steaming Discovery". With that in mind, enforcing `remotePlay.openFirewall` is technically not necessary, since both options can open their respective ports, but opening UDP port 27036 might allow Remote Play clients to be discovered without opening the ports used for actually streaming the game, which might result in confusion - "why does Steam think I can stream, but then when I try to stream gets stuck?"

On the other hand, this being firewall rules, I think I'd be in favor of giving users a choice and documenting this behavior accordingly in both options.

MinerSebas commented:
Strange, I tried it again with UDP port 27036 and TCP port 27040, and now it worked. :man_shrugging:

kira-bruneau commented on nixos/modules/programs/steam.nix:
```diff
@@ -148,6 +156,11 @@ in {
         allowedTCPPorts = [ 27015 ]; # SRCDS Rcon port
         allowedUDPPorts = [ 27015 ]; # Gameplay traffic
       })
+
```
The docs from steam say that 27031-37036 is required for LAN discovery:

https://help.steampowered.com/en/faqs/view/46BD-6BA8-B012-CE43

Since this overlaps with the LAN discovery used for remote play, I think it'd make sense just to separate it into a different `mkIf` block:

```diff
--- a/nixos/modules/programs/steam.nix
+++ b/nixos/modules/programs/steam.nix
@@ -66,6 +66,14 @@ in {
       '';
     };
 
+    localNetworkTransfers.openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Open ports in the firewall for Steam Local Network Game Transfers.
+      '';
+    };
+
     remotePlay.openFirewall = mkOption {
       type = types.bool;
       default = false;
@@ -139,8 +147,16 @@ in {
     ] ++ lib.optional cfg.gamescopeSession.enable steam-gamescope;
 
     networking.firewall = lib.mkMerge [
+      (mkIf cfg.localNetworkTransfers.openFirewall {
+        allowedTCPPorts = [ 27040 ];
+      })
+
       (mkIf cfg.remotePlay.openFirewall {
         allowedTCPPorts = [ 27036 ];
+      })
+
+      (mkIf (cfg.localNetworkTransfers.openFirewall || cfg.remotePlay.openFirewall) {
+        # LAN client discovery
         allowedUDPPortRanges = [ { from = 27031; to = 27036; } ];
       })
```

opl- replied:
I'm not really sure what the convention here is, but my gut is telling me that if less ports need to be open for something to function, then less ports should be open. If the used port change, this can be updated when they do.

If the convention is to follow documentation, then so be it, but I will note that in this instance the documentation was wrong in two different ways: swapped digits 4 and 7, and missing UDP ports, meaning following the docs as they were wouldn't have made this function.

The only reason I know which ports are used in practice is because I checked; the docs were updated following my question to martino, after I had already made the PR.

As for the patch: I do like splitting LAN discovery into its own section. Makes the overlap more obvious.

kira-bruneau replied:
Oh sorry! Yeah I missed your mention of the docs originally being wrong in the PR summary. I was looking into this myself before seeing this PR. I saw the right information there, but didn't realize it was fixed after you reached out 😅.

Yeah that's a good gut reaction. It really doesn't make much sense why they'd need the full range. I assume steam tries to scan it as a fallback in case one of them fails?

I was mainly just thinking that since remote play also uses the same UDP port range, it'd be nice to group them together as generic "lan discovery ports". Maybe we should also reduce remote play to only use one UDP port too?

opl- replied:
We could limit Remote Play ports, obviously assuming it doesn't use those to stream the data and break. I haven't looked into how it works though, and that might be better left for another PR anyway. :P I'd hope they had a reason for listing the other ports as being used by Remote Play though.

I have a mild hunch that since local network transfers use discovery infrastructure used by Remote Play, they might've blindly copied all UDP ports used by Remote Play without fully understanding what they're used for, but that might not be giving them enough credit.

I'm also wondering what ports the Steam process keeps open. It might be a useful clue.

opl- committed:
nixos/steam: add localNetworkTransfers.openFirewall option

Steam local network game transfers require TCP port 27040 to be open:
https://steamcommunity.com/groups/SteamClientBeta/discussions/0/3775742015034590856/#c3827537203130812127

They also require UDP port 27036 to allow discovering peers on the same network before a transfer on port 27040 can be initiated.

Co-authored-by: Kira Bruneau <kira.bruneau@pm.me>

opl- commented:
- Merged client discovery port option into a single `mkIf` as recommended [here](https://github.com/NixOS/nixpkgs/pull/285627#discussion_r1477121437).
- Renamed option from `localNetworkTransfers.openFirewall` to `localNetworkGameTransfers.openFirewall`.
- Rebased onto latest `master`.

kira-bruneau commented:
Thanks @opl-! ❤️
